lib/ts/recipe/passwordless/recipeImplementation.ts

```diff
@@ -0,0 +1,275 @@
+import { RecipeInterface, NormalisedConfig } from "./types";
+import { NormalisedAppInfo } from "../../types";
+import Querier from "../../querier";
+import { PASSWORDLESS_LOGIN_ATTEMPT_INFO_STORAGE_KEY } from "./constants";
+import { getLocalStorage, removeFromLocalStorage, setLocalStorage } from "../../utils";
+
+type FlowType = "USER_INPUT_CODE" | "MAGIC_LINK" | "USER_INPUT_CODE_AND_MAGIC_LINK";
+
+type CreateCodeApiResponse =
+    | {
+          status: "OK";
+          deviceId: string;
+          preAuthSessionId: string;
+          flowType: FlowType;
+      }
+    | {
+          status: "GENERAL_ERROR";
+          message: string;
+      };
+
+type ResendCodeApiResponse = {
+    status: "OK" | "RESTART_FLOW_ERROR";
+};
+
+type ConsumeCodeApiResponse =
+    | {
+          status: "OK";
+          createdUser: boolean;
+          user: {
+              id: string;
+              email?: string;
+              phoneNumber?: string;
+              timeJoined: number;
+          };
+      }
+    | {
+          status: "INCORRECT_USER_INPUT_CODE_ERROR" | "EXPIRED_USER_INPUT_CODE_ERROR";
+          failedCodeInputAttemptCount: number;
+          maximumCodeInputAttempts: number;
+      }
+    | { status: "GENERAL_ERROR"; message: string }
+    | { status: "RESTART_FLOW_ERROR" };
+
+type ExistsAPIResponse = {
+    status: "OK";
+    exists: boolean;
+};
+
+export default function getRecipeImplementation(recipeId: string, appInfo: NormalisedAppInfo): RecipeInterface {
+    const querier = new Querier(recipeId, appInfo);
+
+    return {
+        createCode: async function (
+            input: ({ email: string } | { phoneNumber: string } | { deviceId: string; preAuthSessionId: string }) & {
+                config: NormalisedConfig;
+            }
+        ): Promise<CreateCodeApiResponse> {
+            let bodyObj;
+            if ("email" in input) {
+                const validationRes = await input.config.validateEmailAddress(input.email);
+                if (validationRes !== undefined) {
+                    return {
+                        status: "GENERAL_ERROR",
```

Comment:
If we return a `GENERAL_ERROR`, it shows errors in two places - one below the input field, and one on top